aggregator/src/aggregator.rs

```diff
@@ -2373,24 +2375,23 @@ impl VdafOps {
                     }
 
                     // Write report shares, and ensure this isn't a repeated report aggregation.
-                    try_join_all(report_share_data.iter_mut().map(|rsd| {
+                    let report_aggregations = try_join_all(report_share_data.iter().map(|rsd| {
                         let task = Arc::clone(&task);
```

Comment:
Can we accomplish the same by doing `.iter_mut()` and modifying the elements of `ReportShareData` in place? We would need to at least wrap it in a `std::sync::Mutex` for this to work, but it would avoid copying in all cases and save a bunch of code.

(Just checking if you've considered it, it probably is more involved of a change than I'm thinking).

Reply:
We can't modify `report_share_data`: it's in an `Arc` to avoid copying the `ReportShareData` inside the `Vec`, and `Arc` doesn't allow mutable references. Or, to look at it another way, we can't modify `report_share_data` because we might need the unmodified value if we retry the transaction callback.

Reply:
FWIW I was able to coerce my suggestion into "working" by having `report_share_data` being in an `Arc<tokio::sync::Mutex<_>>`, but it then becomes difficult to take a `.report_aggregation` for use in the `AggregationJobWriter` without `.clone()`ing it--i.e. I landed on needing to pass around a `Cow` anyway. So I think your approach is super-correct :+1:.

Reply:
Yeah -- N.B. the problem with `Arc<Mutex<Vec<T>>>` is that, without a Cow, ultimately we would be modifying the `report_share_data`, which would remain modified in the case of a transaction retry (and on retry, we want to start from the same data as in the previous attempt, to ensure correctness).

As you note, we could do `Arc<Mutex<Vec<Cow<T>>>>`, but at that point, I think we're better off just generating a new `Vec<Cow<T>>` based off of the `report_share_data` since we need to modify the vector anyway.